lib/index.js

```diff
@@ -1,497 +1,730 @@
-  options = bedrock.util.extend({}, options);
-  if(options.filter && typeof options.filter !== 'function') {
-    throw new TypeError('options.filter must be a function.');
-  }
-  /* FIXME: remove filter support and/or improve with include/exclude
-  if(!options.filter) {
-    options.filter = function(field, obj) {
-      return field.indexOf('sys') !== 0;
-    };
+api.update = brCallbackify(async ({actor, id, patch, sequence}) => {
+  assert.string(id, 'id');
+  assert.array(patch, 'patch');
+  assert.number(sequence, 'sequence');
+
+  if(sequence < 0) {
+    throw new TypeError('"sequence" must be a non-negative integer.');
   }
-  */
-  if('changes' in options) {
-    if(!Array.isArray(options.changes)) {
-      if(options.changes && typeof options.changes !== 'object') {
-        throw new TypeError('options.changes must be an object or an array.');
-      }
-      options.changes = [options.changes];
-    }
-    if(typeof identity !== 'string') {
-      throw new TypeError(
-        'When using `options.changes`, `identity` must be a string.');
-    }
-    if(options.changes.length === 0) {
-      // no changes to make, return success
-      return callback();
-    }
-    return _updateIdentity(actor, identity, options, callback);
+
+  const record = await api.get({actor, id});
+  if(record.meta.sequence !== sequence) {
+    return new BedrockError(
+      'Could not update Identity. Record sequence does not match.',
+      'InvalidStateError', {
+        httpStatusCode: 409,
+        public: true,
+        actual: sequence,
+        expected: record.meta.sequence
+      });
   }
-  if(typeof identity !== 'object') {
-    throw new TypeError(
-      'When not specifying `options.changes`, `identity` must be an object.');
+
+  const resource = [id];
+  if(record.identity.owner) {
+    resource.push(record.identity.owner);
   }
+  await _checkPermission({
+    actor,
+    permission: PERMISSIONS.IDENTITY_UPDATE,
+    resource,
+    translate: 'owner'
+  });
 
-  _deprecatedUpdateIdentity(actor, identity, options, callback);
-};
+  const errors = jsonpatch.validate(patch, record.identity);
+  if(errors) {
+    throw new BedrockError(
+      'The given JSON patch is invalid.', 'ValidationError', {
+        httpStatusCode: 400,
+        public: true,
+        patch,
+        errors
+      });
+  }
+
+  // apply patch and validate result
+  const patched = jsonpatch.applyPatch(record.identity, patch).newDocument;
+  await validateInstance(patched, identitySchema);
+  if(patched.memberOf) {
+    await _ensureMembershipValid({actor, identity: patched});
+  }
+
+  const result = await database.collections.identity.update({
+    id: database.hash(id),
+    'meta.sequence': sequence
+  }, {
+    $set: {identity: patched},
+    $inc: {'meta.sequence': 1}
+  }, database.writeOptions);
+
+  if(result.result.n === 0) {
+    return new BedrockError(
+      'Could not update Identity. Record sequence does not match.',
```

Comment:
Could result.n be zero in other cases than this?

Reply:
Not at present.